Redesign topic viewer

aks681 opened the pull request:
## Overview

1. This PR does the following: 

- Redesign the topic viewer according to the mocks [here](https://xd.adobe.com/view/83e97e1d-7820-4387-72b2-8860a2ee8f6d-f5d8/) by @mschanteltc 
- Add a one off job to recompute story summary as 2 new fields (thumbnail_filename and thumbnail_bg_color) have been added along with changing the field node_count to node_titles.
- Modified the practice sessions handler to take a list of subtopics and only ask questions based on skills from those subtopics.
- The story progress section in the mocks hasn't been done as @sseth99 is working on it with the current design.

Screenshots:

**Desktop (1920 x 1080):**

1. Info tab:
![image](https://user-images.githubusercontent.com/22347970/83943605-22398600-a81b-11ea-91ea-23f80983af5e.png)

2. Story tab:
![image](https://user-images.githubusercontent.com/22347970/83943612-2ebdde80-a81b-11ea-9fad-25e74987536c.png)

3. Practice tab:
![image](https://user-images.githubusercontent.com/22347970/83943623-3ed5be00-a81b-11ea-9ae9-e1a667a413e4.png)

4. Subtopics tab:
![image](https://user-images.githubusercontent.com/22347970/83943629-48f7bc80-a81b-11ea-9b3e-770a4b8474ab.png)

**Tablet (768 x 1024) (fairly representative of landscape mobile as well)**

1. Info tab
![image](https://user-images.githubusercontent.com/22347970/83943634-59a83280-a81b-11ea-81da-1ea72aa1f945.png)

2. Story tab
![image](https://user-images.githubusercontent.com/22347970/83943643-64fb5e00-a81b-11ea-8e80-8e9f7e707f8e.png)

3. Practice tab
![image](https://user-images.githubusercontent.com/22347970/83943649-6f1d5c80-a81b-11ea-993f-3ac8f205001f.png)

4. Subtopics tab
![image](https://user-images.githubusercontent.com/22347970/83943654-7ba1b500-a81b-11ea-9764-d2a862c799c7.png)


**Mobile (emulating an iPhone 6)**

1. Info tab:
![image](https://user-images.githubusercontent.com/22347970/83943661-89573a80-a81b-11ea-8928-8e829414938b.png)

2. Story tab:
![image](https://user-images.githubusercontent.com/22347970/83943669-96742980-a81b-11ea-88e0-08a7b6f40305.png)
_(One chapter would be displayed as '1 Chapter' instead of '1 Chapters')_

3. Practice tab:
![image](https://user-images.githubusercontent.com/22347970/83943680-b3a8f800-a81b-11ea-9daa-87a145a6c74a.png)

4. Subtopics tab:
![image](https://user-images.githubusercontent.com/22347970/83943686-bd326000-a81b-11ea-98b2-1a0b35a31839.png)


## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The linter/Karma presubmit checks have passed locally on your machine.
- [ ] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [ ] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

aks681 committed:
Added topic info tab and redesigned tabs

aks681 committed:
Done with info tab

aks681 committed:
started with stories tab

aks681 committed:
revamped story tab

aks681 committed:
disable flag

seanlip commented on core/domain/story_services.py:
```diff
@@ -544,11 +544,12 @@ def compute_summary_of_story(story):
         StorySummary. The computed summary for the given story.
     """
-    story_model_node_count = len(story.story_contents.nodes)
+    story_model_node_titles = [
```
Just to check, a node title can never be None, right? We would error prior to that ever being saved, right?

aks681 replied:
Yeah, chapter titles are asked during creation, so it is never None.

seanlip commented on core/storage/story/gae_models.py:
```diff
@@ -200,7 +200,11 @@ class StorySummaryModel(base_models.BaseModel):
     # model was created).
     story_model_created_on = ndb.DateTimeProperty(required=True, indexed=True)
     # The number of nodes that are part of this story.
```
Docstring is wrong. (When editing it, also mention that it's "in chapter order".)

aks681 replied:
Done

seanlip commented on core/templates/components/summary-tile/story-summary-tile-displayed-chapters-calculator.directive.ts:
```diff
@@ -0,0 +1,46 @@
+ * number of chapters displayed initially.
+ */
+ 
+ /* eslint-disable angular/directive-restrict */
```
Can we not use an attr directive? We are trying to get rid of those from the codebase to make the Angular migration easier.

seanlip replied:
Also, if it helps with getting this PR in, I'm fine with showing a fixed number of chapters for now at the outset. Maybe get guidance from @mschanteltc about what the exact behaviour should be for the number of chapters shown at the outset -- should it be varying based on height or is it OK for us to approximate this by a fixed number?

aks681 replied:
But how can we decide on a fixed number? Since at 1920 x 1080, around 4 chapters can be fit, which would be higher if the display is 4K, but at 1280 x 720 and lower, even 2 can overflow.

So, if we fix it to a low number, then 'Show More' would be shown for high res displays even though the box is barely filled, but if we fix it higher, the box would overflow regardless of the limit for low res displays.

Considering that an 'expand' functionality exists anyway, maybe fixing it to a high value would be better since overflowing for a low res display wouldn't look too bad compared to showing 'Show More' when the box is barely filled.

I tried varying the limit based on window width initially, but figuring out the exact breakpoint would depend on the titles themselves and the width, so that wouldn't work always.

seanlip replied:
I think this is fine for now -- let's fix it to the higher value and could you please start an email thread with @mschanteltc (with screenshots if possible) to ask about what the actual behaviour should be?

aks681 replied:
Ok

seanlip commented on core/templates/components/summary-tile/story-summary-tile.directive.ts:
```diff
@@ -16,34 +16,61 @@
-          ctrl.getStaticImageUrl = function(imagePath) {
-            return UrlInterpolationService.getStaticImageUrl(imagePath);
+          ctrl.showAllChapters = function() {
+            ctrl.stopWatcherOnHeight = true;
```
Not really sure, from the varname, what this means.

aks681 replied:
Oh, this is an unused variable, will remove it.

aks681 replied:
Done

aks681 committed:
made review changes

aks681 commented:
@nithusha21 @ankita240796 @marianazangrossi @kevintab95 PTAL as codeowners.

seanlip commented on core/storage/story/gae_models.py:
```diff
@@ -199,7 +199,7 @@ class StorySummaryModel(base_models.BaseModel):
     # model was created).
     story_model_created_on = ndb.DateTimeProperty(required=True, indexed=True)
-    # The number of nodes that are part of this story.
+    # The title of the nodes in the story, in the same order as present there.
```
Nit: title --> titles

aks681 replied:
Done

jameesjohn commented:
PTAL at the review comments @aks681. Also, take note of the failing checks.

aks681 committed:
Merge remote-tracking branch 'upstream/develop' into topic-viewer

aks681 committed:
Merge remote-tracking branch 'upstream/develop' into topic-viewer

aks681 committed:
mobile topic info tab done

aks681 committed:
story tab mobile done

aks681 committed:
Done with practice sessions tab

aks681 committed:
Finsihed subtopics tab and edits to story tab

aks681 commented:
@seanlip @kevintab95 Updated the description with the new mocks. All the tabs in the topic viewer have been done (including desktop, tablet and mobile view). PTAL!

seanlip commented:
@mschanteltc Please could you take a UI pass at this? It contains all changes for the new topic viewer.

Thanks!

shavavo commented:
![Screen Shot 2020-05-29 at 9 56 10 PM](https://user-images.githubusercontent.com/33111844/83317229-1c2f1e00-a1f9-11ea-990f-fba951440dd4.png)

@aks681  Exception when skill is missing a thumbnail_image

shavavo commented:
@aks681 Hi Akshay, as a heads up, my PR #9397  is using your redesigned subtopic-summary-tile, and has been based off your branch. Please LMK if there are any additional changes to subtopic-summary-tile!

seanlip commented on core/controllers/practice_sessions.py:
```diff
@@ -29,9 +29,15 @@ class PracticeSessionsPage(base.BaseHandler):
+    def get(self, topic_name, selected_skills):
         """Handles GET requests."""
 
+        # Did this to prevent the lint error about unused variable as suggested
```
Hm, let's not introduce a new convention for this. I think you can do "self, unused_topic_name, unused_selected_skills" -- no?

aks681 replied:
Removed it from pathname, reverted this back to just underscore.

seanlip commented on core/controllers/practice_sessions.py:
```diff
@@ -44,24 +50,32 @@ class PracticeSessionsPageDataHandler(base.BaseHandler):
 
     @acl_decorators.can_access_topic_viewer_page
-    def get(self, topic_name):
+    def get(self, topic_name, selected_skills):
```
Probably comma_separated_subtopic_names, since "selected_skills" suggests you're passing in the skill objects themselves. Also, it should be subtopics rather than skills (I know we use the term "skills" in the learner UI but let's be consistent throughout the code in calling them subtopics).

I'm also a bit concerned about the encoding. Subtopic names can contain commas, can't they? Even if not, I think it's worth pretending they can and finding a robust way to handle encoding and decoding so that the data does not get corrupted in transit. (Add backend tests for this case too.)

aks681 replied:
Removed this from path name and using a query string with the stringified array of subtopic titles (as suggested below) as its value.

seanlip commented on core/controllers/practice_sessions.py:
```diff
@@ -44,24 +50,32 @@ class PracticeSessionsPageDataHandler(base.BaseHandler):
         except Exception as e:
             raise self.PageNotFoundException(e)
-        skill_descriptions = {}
+        skill_id_to_description_map = {}
```
skill_ids_to_descriptions

aks681 replied:
Done

seanlip commented on core/templates/components/summary-tile/subtopic-summary-tile.directive.ts:
```diff
@@ -18,34 +18,41 @@
-          ctrl.getSubtopicLink = function() {
-            return UrlInterpolationService.interpolateUrl(
+          ctrl.openSubtopicPage = function() {
+            window.location.href = UrlInterpolationService.interpolateUrl(
```
Should probably use $window or $location?

aks681 replied:
Done

seanlip commented on core/templates/pages/topic-viewer-page/practice-tab/practice-tab.directive.ts:
```diff
@@ -43,12 +43,43 @@ angular.module('oppia').directive('practiceTab', [
           var ctrl = this;
 
           ctrl.newPracticeSession = function() {
```
Start function names with a verb.

aks681 replied:
Done

seanlip commented on core/templates/pages/topic-viewer-page/practice-tab/practice-tab.directive.ts:
```diff
@@ -43,12 +43,43 @@ angular.module('oppia').directive('practiceTab', [
             $window.location.href = practiceSessionsUrl;
           };
+
+          ctrl.disableStartButton = function() {
```
I think this should be isStartButtonDisabled() or shouldStartButtonBeDisabled(), since it returns a boolean rather than performs an action.

aks681 replied:
Done

seanlip commented on core/templates/pages/topic-viewer-page/subtopics-list/subtopics-list.directive.html:
```diff
@@ -1,107 +1,55 @@
+<div class="oppia-subtopics-list-container">
+  <md-card class="oppia-page-card oppia-long-text">
+    <h2 class="tab-title">Study these Skills</h2>
+    <span class="subtopic-subheading"> Study skills by going over important concepts. </span>
```
Text here, and elsewhere, needs i18n?

aks681 replied:
Done, added translation everywhere.

seanlip commented on core/templates/services/contextual/url.service.ts:
```diff
@@ -85,6 +85,14 @@ export class UrlService {
     throw new Error('Invalid URL for topic');
   }
 
+  getSelectedSubtopicsFromUrl(): string {
```
Don't forget to add tests :)

aks681 replied:
Done

seanlip commented on core/templates/pages/topic-viewer-page/topic-viewer-page.directive.html:
```diff
@@ -1,54 +1,94 @@
+      <practice-tab subtopics-list="$ctrl.subtopics"
+                    topic-name="$ctrl.topicName">
+      </practice-tab>
+      <!-- <skills-mastery-list degrees-of-mastery="$ctrl.degreesOfMastery" skill-descriptions="$ctrl.skillDescriptions"></skills-mastery-list> -->
```
Don't commit commented-out code.

aks681 replied:
Ah, I forgot to add a comment in the PR thread here for this. So, I wanted to ask, this used to be shown in the topic viewer but the new mocks doesn't have this displayed, so should I remove it completely for now and add it later on?

aks681 replied:
I've deleted this for now.

seanlip commented on core/templates/pages/topic-viewer-page/practice-tab/practice-tab.directive.ts:
```diff
@@ -43,12 +43,43 @@ angular.module('oppia').directive('practiceTab', [
-                topic_name: ctrl.getTopicName()
+                topic_name: ctrl.getTopicName(),
+                // These are actually the selected subtopics, but named so
+                // because this will be shown in the URL and for the learner,
```
I think it's ok to still call them subtopics; learners won't pay attention to long urls that much.

aks681 replied:
Done

seanlip commented on core/templates/pages/topic-viewer-page/practice-tab/practice-tab.directive.ts:
```diff
@@ -43,12 +43,43 @@ angular.module('oppia').directive('practiceTab', [
+            for (var idx in ctrl.selectedSubtopicIndices) {
+              if (ctrl.selectedSubtopicIndices[idx]) {
+                commaSeparatedSelectedSubtopics += (
+                  ctrl.availableSubtopics[idx].getTitle() + ',');
```
May need to be careful about encoding here. (Perhaps create the list and encode as json string?)

aks681 replied:
Done, made it a query string with stringified array, read using json.loads in the backend.

seanlip reviewed (changes requested):
Thanks @aks681, took a pass!

mschanteltc commented:
@aks681 Each tab is looking great so far! Here are my comments for each page:

**General**
- For the green color seen in the header, tab menu, and "View all" button, use #00645C because it meets WCAG AAA contrast requirements
- Use realistic content for the text and images. Use this [document](https://docs.google.com/spreadsheets/d/1j2XfPz2Y_q9qO44ZZLFPkjWF2_8Y_Ally-Dqe0NQTy4/edit#gid=0) for text and this [image folder](https://drive.google.com/drive/u/1/folders/1qb3m0BH9PXvDhoMjA2hVPnguoh-PGSHk) for thumbnails.

**[Info Tab](https://xd.adobe.com/view/83e97e1d-7820-4387-72b2-8860a2ee8f6d-f5d8/)**
- There should be a "Prerequisites" portion that lists what math topics the user must have completed before doing this topic
- Mobile: The numbers in the statistics and description (Stories, Lessons, Skills) don't look exactly centered

**[Story Tab](https://xd.adobe.com/view/83e97e1d-7820-4387-72b2-8860a2ee8f6d-f5d8/screen/ac9d4005-91c9-419d-8823-0153cf622f18/Lessons-Tab)**
- The Chapter List should have even top and bottom spacing from the line dividers
- "View All" should be centered respective of the chapter list box
- Can we get an image of how a "complete" Chapter looks like with the check mark next to it?
- Let have Chapter 1 look "completed" and lower the Chapter 3 opacity because the user cannot progress
- Add the percentage progress indicator left of the Story name
- Make sure the number and noun make grammatical sense ("1 Chapters" -> "1 Chapter")
- Mobile: thumbnail image's bottom left and right corners should be 90 degree angles, not curved
- Mobile: Also missing percentage indicator

**[Practice Tab](https://xd.adobe.com/view/83e97e1d-7820-4387-72b2-8860a2ee8f6d-f5d8/screen/ec2ef647-dfd4-4135-a30d-a74665ae5ecb/Practice-Tab)**
- "Select skills that you would like to practice" should use Roboto font
- The white box should be narrower compared to that of Info, Lessons, and Revision to make the checklist seem more center balanced while still being left aligned 
- "Start" button should use Roboto font
- The checkbox be top aligned with the text, not centered

**[Revision Tab](https://xd.adobe.com/view/83e97e1d-7820-4387-72b2-8860a2ee8f6d-f5d8/screen/76310d94-e8f7-484e-ac44-96a83bf7a8cf/Revision-Tab)**
- Card names should use medium weight Roboto font
- "Study skills by going over important concepts" should use Roboto font
- The orange color used is #BE563C

aks681 commented:
@shavavo By skill, do you mean subtopics? Since skills don't have thumbnails. If so, we do actually require all subtopics to have a thumbnail before publishing, so that wouldn't be an issue for learners.
Not sure what we can do in the preview tab for such subtopics, other than hide them completely.
Also, I would be editing the subtopic summary tile when addressing the review comments, so just a heads up!

@mschanteltc Ok, will do the changes! As for the story progression, that is being tackled by @sseth99 for the old design, so I thought I'll incorporate that once she's done with it, is that fine?

seanlip commented:
Hi @aks681 -- re progress, just confirming that should be OK. Thanks!

aks681 committed:
Addressed sean's comments

aks681 commented:
@mschanteltc We currently don't have the models to show 'Prerequisite Topics' for a topic, so is it fine to do that in a separate PR? Also, how would the info tab be like, if there are prerequisite topics?
The mocks that you had sent had no prerequisite topics in them, so what would be shown there if there were some, is it just a bulleted list of names?

seanlip commented:
For prerequisite topics, it's fine to do that in a separate PR (please file an issue).

(I'll defer to @mschantelc re the other questions about what the mocks should look like in that case. Thanks!)

shavavo commented on core/domain/story_domain.py:
```diff
@@ -1369,15 +1374,31 @@ def validate(self):
-                'Expected node_count to be an int, received \'%s\'' % (
-                    self.node_count))
+                'Expected node_titles to be a list, received \'%s\'' % (
+                    self.node_titles))
```
Nit pick: perhaps log type(self.node_titles) here

shavavo replied:
Here and elsewhere

aks681 replied:
This is how validations are done elsewhere in the codebase as well, so not sure if we should change it here. Also, logging the whole thing would help in identifying the type as well, right?

shavavo replied:
You're right, I guess logging it should be enough in most cases

shavavo replied:
Done.

shavavo commented on core/domain/story_domain.py:
```diff
@@ -1369,15 +1374,31 @@ def validate(self):
+            if not isinstance(title, python_utils.BASESTRING):
+                raise utils.ValidationError(
+                    'Expected each chapter title to be a string, received %s'
+                    % title)
```
Ditto

aks681 replied:
Ditto above

shavavo replied:
Done.

shavavo commented on core/templates/components/summary-tile/story-summary-tile.directive.html:
```diff
@@ -1,42 +1,141 @@
+  }
+
+  story-summary-tile a {
+    color: black;
```
Replace with #000000

aks681 replied:
Done

shavavo commented on core/templates/components/summary-tile/subtopic-summary-tile.directive.html:
```diff
@@ -1,35 +1,84 @@
+  }
+
+  subtopic-summary-tile .subtopic-thumbnail {
+    border: 1px sold white;
```
replace white with hex

aks681 replied:
Done

shavavo commented on core/templates/pages/topic-viewer-page/topic-viewer-page.controller.ts:
```diff
@@ -54,28 +55,40 @@ angular.module('oppia').directive('topicViewerPage', [
                 LoaderService.hideLoadingScreen();
                 ctrl.trainTabShouldBeDisplayed = (
                   readOnlyTopic.getTrainTabShouldBeDisplayed());
+                ctrl.topicIsLoading = false;
```
considering moving this above LoaderService.hideLoadingScreen(); on line 88 if it effects layout

aks681 replied:
Done

shavavo commented on core/templates/pages/topic-viewer-page/topic-viewer-page.directive.html:
```diff
@@ -61,23 +101,32 @@
   .oppia-topic-viewer-tabs-container .topic-viewer-tabs .topic-viewer-tabs-text {
-    color: #009688;
-    font-size: 1em;
+    color: white;
```
ditto w/ color

aks681 replied:
Done

shavavo commented on core/templates/pages/topic-viewer-page/topic-viewer-page.directive.html:
```diff
@@ -61,23 +101,32 @@
-    color: #009688;
-    margin-left: 7.5px;
+    color: white;
+    border-bottom: 4px solid white;
```
ditto

aks681 replied:
Done

shavavo commented on core/templates/pages/topic-viewer-page/topic-viewer-page.directive.html:
```diff
@@ -61,23 +101,32 @@
 
   .oppia-topic-viewer-tabs-container .topic-viewer-tabs .topic-viewer-tabs-text:hover {
-    color: #009688;
+    color: white;
```
ditto

aks681 replied:
Done

shavavo commented on core/templates/pages/topic-viewer-page/topic-viewer-page.directive.html:
```diff
@@ -144,12 +207,9 @@
-    background-color: #e7e7e2;
-    border-top: 4px solid #009688;
-    color: #009688;
+    border-bottom: 4px solid white;
```
ditto

aks681 replied:
Done

aks681 committed:
addressed UI comments and other changes

aks681 committed:
Merge remote-tracking branch 'upstream/develop' into topic-viewer

aks681 committed:
minor change

aks681 commented:
@seanlip @mschanteltc @shavavo PTAL. (Also @shavavo, I have change the subtopic summary tile slightly).

shavavo commented:
LGTM Thanks @aks681

mschanteltc commented:
> @mschanteltc We currently don't have the models to show 'Prerequisite Topics' for a topic, so is it fine to do that in a separate PR? Also, how would the info tab be like, if there are prerequisite topics?
> The mocks that you had sent had no prerequisite topics in them, so what would be shown there if there were some, is it just a bulleted list of names?

Great question Akshay. To view the [mock](https://xd.adobe.com/view/83e97e1d-7820-4387-72b2-8860a2ee8f6d-f5d8/) of a Topic with prerequisites, tap on the word "Prerequisites." I added more info in the note, but let me know if you have any more questions.

seanlip commented on core/controllers/practice_sessions.py:
```diff
@@ -52,16 +54,22 @@ def get(self, topic_name):
         # Topic cannot be None as an exception will be thrown from its decorator
         # if so.
         topic = topic_fetchers.get_topic_by_name(topic_name)
+        subtopic_titles_list = self.request.get('selected_subtopics')
```
subtopic_titles_json perhaps? (and change the url param name to include "json" too)

aks681 replied:
I am still stringifying a js array and not a json object, so is 'json' needed in name?

seanlip replied:
Yes because this is a string in json format, right? Only once it stops being in json format and is turned into a list, should you call it a list.

aks681 replied:
Renamed it, used subtopic IDs (which are numbers) and made it a comma separated string.

seanlip commented on core/controllers/practice_sessions.py:
```diff
@@ -52,16 +54,22 @@ def get(self, topic_name):
         # if so.
         topic = topic_fetchers.get_topic_by_name(topic_name)
+        subtopic_titles_list = self.request.get('selected_subtopics')
+        selected_subtopic_titles = json.loads(subtopic_titles_list)
```
try/except here, return 400/404 if invalid format (remember, don't trust any input that comes in from the outside)

aks681 replied:
Using split now to get IDs from string, in which case even if it is not a correct string,it's fine as it would just be considered as an invalid subtopic ID and never matched.

seanlip commented on core/controllers/practice_sessions.py:
```diff
@@ -52,16 +54,22 @@ def get(self, topic_name):
+        selected_skill_ids = []
+        for subtopic in topic.subtopics:
+            if subtopic.title in selected_subtopic_titles:
+                selected_skill_ids.extend(subtopic.skill_ids)
```
Else raise 400 (if there's some invalid item in the list)? Or should we not do that?

aks681 replied:
I think it's fine to just ignore those? If some subtopics were deleted after the learner page was loaded, then this would throw an error, but it shouldn't, right?

seanlip replied:
OK, sounds reasonable! Make sure you have a backend test for this case. Also, here, perhaps add a comment with some explanation of the desired behaviour and the rationale for it.

seanlip commented on core/controllers/practice_sessions_test.py:
```diff
@@ -132,30 +135,28 @@ def test_get_fails_when_skill_ids_dont_exist(self):
         with self.swap(constants, 'ENABLE_NEW_STRUCTURE_PLAYERS', True):
             self.get_json(
-                '%s/%s' % (
+                '%s/%s?selected_subtopics=["subtopic_name"]' % (
```
Just a question about this. Do we have the concept of subtopic ids? If so, any reason not to use those instead? It might look better in the url than a list (and if the ids are from a restricted charset then the issues with comma-separation might go away).

aks681 replied:
Yeah, changed it to use comma separated subtopic IDs

seanlip commented on core/templates/pages/practice-session-page/practice-session-page.constants.ts:
```diff
@@ -20,9 +20,11 @@ export class PracticeSessionPageConstants {
   public static PRACTICE_SESSIONS_DATA_URL =
-    '/practice_session/data/<topic_name>';
+    '/practice_session/data/<topic_name>?' +
+    'selected_subtopics=<subtopic_titles_list>';
```
It's not a list, it's a json string, so that should be reflected in the param name.

aks681 replied:
Renamed it

seanlip commented on core/templates/pages/topic-viewer-page/info-tab/topic-info-tab.directive.ts:
```diff
@@ -0,0 +1,50 @@
+          var ctrl = this;
+
+          ctrl.$onInit = function() {
+            ctrl.isMobile = false;
```
screenHasSmallWidth might be a better name -- "isMobile" suggests an actual mobile device which may not be the case.

aks681 replied:
Done

seanlip commented on core/templates/pages/topic-viewer-page/practice-tab/practice-tab.directive.html:
```diff
@@ -1,31 +1,68 @@
-  .oppia-practice-tab-container {
-    margin-top: 20px;
+  practice-tab .oppia-long-text {
+    /* Since the same class is used in all the other tabs, it is defined in topic-viewer-page.directive.html,
```
Wait, I don't understand. Why is the practice-tab class used in all the other tabs? 

And if you're referring to .oppia-long-text, why isn't the additional practice-tab qualifier sufficient to special-case it? Would adding additional CSS qualifiers help (like an intermediate .oppia-practice-tab-container)? You'll want enough qualifiers to ensure that the rules defined here are scoped to this template, and if you need other rules to be defined outside, define them outside using different selectors. Otherwise it starts getting messy.

aks681 replied:
Done, removed !important

seanlip commented on core/templates/pages/topic-viewer-page/practice-tab/practice-tab.directive.html:
```diff
@@ -1,31 +1,68 @@
+
+  @media (max-width: 1024px) {
+    practice-tab .oppia-long-text {
+      width: 60vw !important;
```
Same thing here and below, try to avoid !important

aks681 replied:
Done

seanlip reviewed (changes requested):
Thanks @aks681! Took another pass.

aks681 committed:
Merge remote-tracking branch 'upstream/develop' into topic-viewer

aks681 committed:
addressed comments

aks681 committed:
added directive test

aks681 committed:
fixed typescript error

aks681 committed:
fixed backend errors

nishantwrp reviewed (approved):
lgtm as a codeowner!

seanlip reviewed (approved):
LGTM from my end (just one minor comment). Thanks @aks681!

seanlip commented:
Hi @kevintab95 @nithusha21 @ankita240796 @DubeySandeep @marianazangrossi -- could you PTAL as codeowners? This is a bit time-sensitive since the PR needs to go in for the release.

@aks681 please note that the e2e checks on Travis are failing. (Might want to loop in @U8NWXD @DubeySandeep @nithusha21 if this is a flake.)

kevintab95 reviewed (approved):
Thanks @aks681! LGTM for code-owner files.

mschanteltc commented:
Thank you for the update. Here are my comments:

**General**:
- Can we remove the drop shadow from the nav bar and use a drop shadow for the tab menu instead? I know that this might be difficult to do since the navbar is a global component, so maybe we can have the tab menu "layered above" the nav bar so that we don't see the nav bar's drop shadow. 
- The tab menu options can be closer together. Can we use the dimensions used in the mocks? These would be a fixed size and spacing from each other if the window were to change size (until it a breakpoint).
- Web: The white container seems wider than the mocks, making some of the components within this container wider than 

**Info Tab**:
- Statistics need to be grammatically correct. Example: "1 Stories" needs to be "1 Story"
- Web: There should be equal amount of padding on the left and right. Reduce the right padding. 
- Tablet: Looking great!
- Mobile: Add more top/right/bottom/left padding to the container. The mocks use 32px/24px/64px/24px padding. The statistics container(i.e. "1 Story" gray background) would need to  have a smaller width so it doesn't look cramped. 

**Story Tab**: 
- Make sure the chapter list shows reflects the user's progress. The chapters in a chapter list with multiple chapters should either have a checkmark or reduced opacity.
- Web: The chapter list container (as well as the "Lessons" container) should be smaller in width. Also add more bottom padding
- Tablet: "Jamie goes to the arcade" font looks like it is bold, when it should use a medium weight. 
- Mobile: Is there any way we can keep the "__ Chapters" together on the same line? This would apply to the other device formats 

**Practice Tab**:
- Web: The container could be even smaller in width. Although we expect majority of the content to be on the left side of the container, there seems to be an jarring amount of white space on the right side. 
- Tablet: Add more bottom padding to the container. 
- Mobile: Looking good! Might need to change padding to keep it consistent with the other tabs.

**Revision Tab**:
- Web: Add more bottom padding. (keep the padding consistent to all white containers of each tab)
- Mobile: reduce the column spacing between the cards.

mschanteltc commented:
**Skill Card**:
- The orange color used is #BE563C (current mock looks darker)
- Correction to my previous comment: names should use a bold weight, not medium
- Try to match the dimensions, padding, and font sizes with the [mock's](https://xd.adobe.com/view/83e97e1d-7820-4387-72b2-8860a2ee8f6d-f5d8/screen/76310d94-e8f7-484e-ac44-96a83bf7a8cf/Revision-Tab)

aks681 committed:
addressed code comments

aks681 committed:
Merge remote-tracking branch 'upstream/develop' into topic-viewer

nithusha21 reviewed (approved):
LGTM for codeowner files (assets/)

aks681 committed:
addressed ui changes

ankita240796 reviewed (approved):
LGTM from codeowner's perspective, Thanks @aks681!

aks681 committed:
addressed ui comments

aks681 commented:
@mschanteltc Addressed the UI comments, though in the font size we are working with here, I couldn't vary the font-weight from regular -> medium -> bold in CSS. There were only 2 steps and it changed directly from regular to bold, hence why the titles may look slightly bolder. Having it regular made it too thin. 
Also:
- The orange color used is #BE563C, just an issue with the screenshot itself.
- Kept the width of the white box fixed, as per discussion, in the web view to 704px.

DubeySandeep commented on main.py:
```diff
@@ -263,10 +263,12 @@ def ui_access_wrapper(self, *args, **kwargs):
     get_redirect_route(
-        r'%s/<topic_name>' % feconf.PRACTICE_SESSION_URL_PREFIX,
+        r'%s/<topic_name>' %
+        feconf.PRACTICE_SESSION_URL_PREFIX,
```
Any specific reasonf or this change?

aks681 replied:
Oh, no. I had added another field in the middle, but removed that later on, will revert this back.

aks681 replied:
@shavavo has reverted these in #9397

DubeySandeep commented on main.py:
```diff
@@ -263,10 +263,12 @@ def ui_access_wrapper(self, *args, **kwargs):
     get_redirect_route(
-        r'%s/<topic_name>' % feconf.PRACTICE_SESSION_DATA_URL_PREFIX,
+        r'%s/<topic_name>' %
+        feconf.PRACTICE_SESSION_DATA_URL_PREFIX,
```
Ditto

aks681 replied:
Ditto

DubeySandeep reviewed (commented):
LGTM, just a minor comment!

marianadasilvadev reviewed (approved):
LGTM as codeowner perspective

mschanteltc commented:
**General**:
- Reduce the opacity of all the tabs except the one that is not selected. I would say reduce it to 70% opacity.
- For the Tab names, use Roboto font.


**Info Tab**:
- Looks great to me! 

**Lessons Tab**:
- Lets add more top/bottom spacing to the Chapter List names and lines

**Practice Tab**:
- Also looks great to me!

**Revision Tab**:
- Add more top/bottom margins to the cards (ex: In Web view, "Place names and their values" and Rounding numbers" having small spacing between each other. Try to make that spacing equal to the left/right margins)

seanlip commented:
@aks681 can you file an issue / give an update on this thread about the response to @mschanteltc's latest comments please? Thanks!
